core/rio/jsonld/src/main/java/org/eclipse/rdf4j/rio/ndjsonld/NDJSONLDParser.java

```diff
@@ -0,0 +1,97 @@
+import org.eclipse.rdf4j.rio.jsonld.JSONLDParser;
+
+import java.io.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+public class NDJSONLDParser extends AbstractRDFParser implements RDFParser {
+
+    private JSONLDParser getJsonLDParser() {
+        JSONLDParser jsonldParserDelegate = new JSONLDParser();
+        jsonldParserDelegate.setRDFHandler(new RDFHandlerWrapper(getRDFHandler()) {
+            boolean startedRDF;
+
+            // Reuse the same parser for each line parse, but we cannot start the RDF multiple times, so call startRDF only the first time.
+            // An alternative approach would be to to create a new JSONLDParser for each row. 
+            @Override
+            public void startRDF() throws RDFHandlerException {
+                if (!startedRDF) {
+                    startedRDF = true;
+                    super.startRDF();
+                }
+            }
+        });
+        jsonldParserDelegate.setParserConfig(getParserConfig());
+        jsonldParserDelegate.setParseErrorListener(getParseErrorListener());
+        if (this.valueFactory != null) {
+            jsonldParserDelegate.setValueFactory(this.valueFactory);
+
+        }
+        return jsonldParserDelegate;
+    }
+
+    /**
+     * Default constructor
+     */
+    public NDJSONLDParser() {
+        super();
+    }
+
+    /**
+     * Creates a Sesame NDJSONLD Parser using the given {@link ValueFactory} to create new {@link Value}s.
+     *
+     * @param valueFactory The ValueFactory to use
+     */
+    public NDJSONLDParser(final ValueFactory valueFactory) {
+        super(valueFactory);
+    }
+
+    @Override
+    public RDFFormat getRDFFormat() {
+        return RDFFormat.NDJSONLD;
+    }
+
+    @Override
+    public void parse(InputStream in, String baseURI) throws RDFParseException, RDFHandlerException, IOException {
+        try (InputStreamReader isr = new InputStreamReader(in)) {
+            parseInternal(isr, baseURI);
+        }
+    }
+
+    @Override
+    public void parse(Reader reader, String baseURI) throws RDFParseException, RDFHandlerException {
+        parseInternal(reader, baseURI);
+    }
+
+
+    private void parseInternal(Reader reader, String baseURI)
+            throws RDFParseException, RDFHandlerException {
+
+        clear();
+        JSONLDParser jsonldParserDelegate = getJsonLDParser();
+        final AtomicInteger lineCount = new AtomicInteger(0);
+
+        try (Stream<String> lines = new BufferedReader(reader).lines()) {
+            lines.forEach(line -> {
+                try {
+                    lineCount.incrementAndGet();
+                    if (!line.trim().isEmpty()) {
+                        jsonldParserDelegate.parse(new ByteArrayInputStream(line.getBytes()), baseURI);
+                    }
+                } catch (IOException e) {
+                    throw new RDFParseException(String.format("Could not parse a line {}. Error: {}", lineCount.get() - 1), e);
```

Comment:
I don't think this String.format() is valid. It needs %-something instead of {} and you passed only 1 argument. I don't think you should catch IOException either, so best just to leave it propagate.
However, it would be nice to catch a parser exception from the JSON-LD parser and add the line number to the message. Something like
```
} catch (RDFParseException e) {
    new RDFParseException(String.format("Could not parse JSON-LD document at NDJSON-LD line %d: %s", lineCount.get(), e.getMessage), e);
}
```

Reply:
Thanks for spotting the wrong String.format, I catch and re-throw the IOException because it cannot be propagated from the Lambda function. And I prefer using a stream to read the lines rather than filling them all in memory. Any other suggestions on the IOException handling?

Reply:
Left the IOException be propagated